src/analyser.cpp

```diff
@@ -1117,12 +1243,785 @@ void Analyser::AnalyserImpl::analyseEquationAst(const AnalyserEquationAstPtr &as
+    //  - Qualifier elements:
+    //     - 'bvar': a bounding variable can have any unit.
+    //     - 'degree': a degree should be dimensionless.
+    //     - 'logbase': a base should be dimensionless.
+    //  - Trigonometric operators ('sin', 'cos', 'tan', etc.): the argument
+    //    should be dimensionless.
+    //  - Mathematical and logical constants ('pi', 'exponentiale',
+    //    'notanumber','infinity', 'true' and 'false'): those constants are
+    //    dimensionless.
+
+    // Make sure that we have an AST to analyse.
+
+    if (ast == nullptr) {
+        unitsMaps = {};
+        userUnitsMaps = {};
+        unitsMultipliers = {};
+
+        return;
+    }
+
+    // Check whether we are dealing with a CI/CN element and, if so, retrieve
+    // both its units maps and multipliers.
+
+    if ((ast->mPimpl->mType == AnalyserEquationAst::Type::CI)
+        || (ast->mPimpl->mType == AnalyserEquationAst::Type::CN)) {
+        auto units = mCiCnUnits[ast].lock();
+        auto model = owningModel(units);
+
+        defaultUnitsMapsAndMultipliers(unitsMaps, userUnitsMaps, unitsMultipliers);
+
+        for (auto &unitsMap : unitsMaps) {
+            updateUnitsMap(model, units->name(), unitsMap);
+        }
+
+        for (auto &userUnitsMap : userUnitsMaps) {
+            updateUnitsMap(model, units->name(), userUnitsMap, true);
+        }
+
+        for (auto &unitsMultiplier : unitsMultipliers) {
+            updateUnitsMultiplier(model, units->name(), unitsMultiplier);
+        }
+
+        return;
+    }
+
+    // Check the left and right children.
+
+    auto oldNbOfIssueDescriptions = issueDescriptions.size();
+    UnitsMaps rightUnitsMaps;
+    UnitsMaps rightUserUnitsMaps;
+    UnitsMultipliers rightUnitsMultipliers;
+
+    analyseEquationUnits(ast->mPimpl->mOwnedLeftChild, unitsMaps, userUnitsMaps, unitsMultipliers, issueDescriptions);
+    analyseEquationUnits(ast->mPimpl->mOwnedRightChild, rightUnitsMaps, rightUserUnitsMaps, rightUnitsMultipliers, issueDescriptions);
+
+    if ((ast->mPimpl->mType == libcellml::AnalyserEquationAst::Type::ASSIGNMENT)
+        || (ast->mPimpl->mType == libcellml::AnalyserEquationAst::Type::EQ)
+        || (ast->mPimpl->mType == libcellml::AnalyserEquationAst::Type::NEQ)
+        || (ast->mPimpl->mType == libcellml::AnalyserEquationAst::Type::LT)
+        || (ast->mPimpl->mType == libcellml::AnalyserEquationAst::Type::LEQ)
+        || (ast->mPimpl->mType == libcellml::AnalyserEquationAst::Type::GT)
+        || (ast->mPimpl->mType == libcellml::AnalyserEquationAst::Type::GEQ)
+        || (ast->mPimpl->mType == libcellml::AnalyserEquationAst::Type::PLUS)
+        || (ast->mPimpl->mType == libcellml::AnalyserEquationAst::Type::MINUS)
+        || (ast->mPimpl->mType == libcellml::AnalyserEquationAst::Type::MIN)
+        || (ast->mPimpl->mType == libcellml::AnalyserEquationAst::Type::MAX)
+        || (ast->mPimpl->mType == libcellml::AnalyserEquationAst::Type::REM)) {
```

Comment:
Maybe we should add all of these types to a set and then ask if the type is in the set.

Reply:
FWIW, I have implemented the `set` approach and did a quick benchmark (by checking units 1,000 times for each equation in our units checking tests) and the `set` approach is 0.3% faster. So, both approaches are equivalent. Yet, I prefer the `||` approach since one can then see all the operators at a glance, making the code easier to maintain, understand, etc.